applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,576 @@
+                    if self.relaxation_coefficients[index] < 1.0:
+                        self.relaxation_coefficients[index] = min(self.relaxation_coefficients[index] + 0.02, 1.0)
+                        self.direction_has_changed = True
+                    elif self.correction_coefficients[index] < 2.0:
+                        self.correction_coefficients[index] = min (self.correction_coefficients[index] + 0.02, 2.0)
+                        self.direction_has_changed = True
+                KM.Logger.PrintInfo("Constraint ", identifier, "W_R, W_C = ", self.relaxation_coefficients[index], self.correction_coefficients[index])
+
+
+
+    # --------------------------------------------------------------------------
+    def __LineSearch(self):
+        KM.Logger.PrintInfo("Line Search ...")
+        if self.line_search_type == "manual_stepping":
+            self.__manualStep()
+        elif self.line_search_type == "QNBB_method":
+            if self.optimization_iteration == 1:
+                self.step_size /= 5
+                self.__manualStep()
+                self.step_size *= 5
+            else:
+                self.__QNBBStep()
+
+
+    def __manualStep(self):
+        step_norm = self.optimization_utilities.ComputeMaxNormOfNodalVariable(self.design_surface, KSO.CONTROL_POINT_UPDATE)
+        if abs(step_norm) > 1e-10:
+            step = KM.Vector()
+            self.optimization_utilities.AssembleVector(self.design_surface, step, KSO.CONTROL_POINT_UPDATE)
+            step *= 1.0 / step_norm
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, step, KSO.SEARCH_DIRECTION)
+            step *= self.step_size
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, step, KSO.CONTROL_POINT_UPDATE)
+
+    def __QNBBStep(self):
+        s_norm = self.optimization_utilities.ComputeMaxNormOfNodalVariable(self.design_surface, KSO.SEARCH_DIRECTION)
+        print("before s_norm", s_norm)
+        if abs(s_norm) > 1e-10:
+            s = KM.Vector()
+            self.optimization_utilities.AssembleVector(self.design_surface, s, KSO.SEARCH_DIRECTION)
+            s /= s_norm
+            self.optimization_utilities.AssignVectorToVariable(self.design_surface, s, KSO.SEARCH_DIRECTION)
+
+            for index, node in enumerate(self.design_surface.Nodes):
+                i = index * 3
+                y_i = np.array(self.prev_s[i: i+3]) - np.array(s[i: i+3])
+                d_i = np.array(self.d[i:i+3])
+                if np.dot(y_i, y_i) < 1e-9:
+                    step_i = self.step_size
+                else:
+                    step_i = abs(np.dot(d_i, y_i) / np.dot(y_i, y_i))
+                if step_i > self.step_size:
+                    step_i = self.step_size
+                node.SetSolutionStepValue(KSO.INV_HESSIAN, step_i)
+                s[i] = s[i] * step_i
+                s[i+1] = s[i+1] * step_i
+                s[i+2] = s[i+2] * step_i
```

Comment:
```suggestion
                s[i] *= step_i
                s[i+1] *= step_i
                s[i+2] *= step_i
```
not sure if this works for Vector entries